src/ucp/core/ucp_am.c

```diff
@@ -1511,24 +1560,20 @@ ucs_status_t ucp_am_rndv_process_rts(void *arg, void *data, size_t length,
                                           &param.reply_ep);
     status          = am_cb->cb(am_cb->context, hdr, rts->am.header_length,
                                 desc + 1, rts->super.size, &param);
-    if ((status == UCS_INPROGRESS) ||
-        (desc->flags & UCP_RECV_DESC_FLAG_RNDV_STARTED)) {
-        if (desc->flags & UCP_RECV_DESC_FLAG_COMPLETED) {
-            /* User initiated rendezvous receive in the callback and it is
-             * already completed. No need to save the descriptor for further use
-             */
-            goto out;
-        }
-
+    if (ucp_am_rdesc_in_progress(desc, status)) {
         /* User either wants to save descriptor for later use or initiated
          * rendezvous receive (by ucp_am_recv_data_nbx) in the callback. */
         ucs_assert(!UCS_STATUS_IS_ERR(status));
 
-        /* Set this flag after the callback invocation to distiguish the cases
-         * when ucp_am_recv_data_nbx is called inside the callback or not.
-         */
-        desc->flags |= UCP_RECV_DESC_FLAG_RNDV;
+        desc->flags |= UCP_RECV_DESC_FLAG_INTIALIZED;
+
         return desc_status;
+    } else if (desc->flags & UCP_RECV_DESC_FLAG_COMPLETED) {
```

Comment:
we already checked COMPLETED inside ucp_am_rdesc_in_progress, can we avoid extra branch?

Reply:
we have to distinguish 2 different cases, hard to do that without a branch

Reply:
removed `UCP_RECV_DESC_FLAG_COMPLETED`